[UnifiedPDF] Start creating context menu in UnifiedPDFPlugin.

sammygill opened the pull request:
#### 6a92aded2a217de967647f6a04139b078abf36c5
<pre>
[UnifiedPDF] Start creating context menu in UnifiedPDFPlugin.
<a href="https://bugs.webkit.org/show_bug.cgi?id=265605">https://bugs.webkit.org/show_bug.cgi?id=265605</a>
<a href="https://rdar.apple.com/119001947">rdar://119001947</a>

Reviewed by Simon Fraser.

UnifiedPDF will need to start creating its own context menu so we can
lay the initial groundwork by creating a context menu with a single
item. This will hopefully allow us to easily add additional items in
a subsequent patch without worrying about the architecture too much.
In this patch the item we will use will be the &quot;Open with Preview,&quot;
option.

UnifiedPDFPlugin will create its own PDFContextMenu with items and
send this over to the web page proxy. Each item will contain a tag that
is a ContextMenuAction that will be used to indicate what type of action
to perform once we get the result back.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::contextMenuItemPDFOpenWithPreview):
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::createContextMenu const):
(WebKit::UnifiedPDFPlugin::handleContextMenuEvent):
(WebKit::UnifiedPDFPlugin::performContextMenuAction const):

Canonical link: <a href="https://commits.webkit.org/271429@main">https://commits.webkit.org/271429@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b967f0f6bd485c782d36795b5cfed16cca3e6bbd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28338 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6982 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29719 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30865 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/25800 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28835 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9059 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4353 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26071 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28607 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5741 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24382 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5019 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5119 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/25381 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/31551 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25942 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25816 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31425 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5090 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3270 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29183 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6685 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6799 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5540 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5608 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -70,11 +71,12 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     bool handleEditingCommand(StringView commandName) override;
     bool isEditingCommandEnabled(StringView commandName) override;
 
+    NSMenu* createContextMenu(const WebMouseEvent&, WebCore::UserInterfaceLayoutDirection) final;
```
ObjC so `NSMenu *`

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -126,6 +128,8 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
 
     WebCore::ScrollingNodeID m_scrollingNodeID { 0 };
+
+    RetainPtr<WKPDFContextMenuHandler> m_wkPdfContextMenuHandler;
```
Maybe just `m_contextMenuHandler`

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -28,6 +28,7 @@
 
 #if ENABLE(UNIFIED_PDF)
 
+#include "PdfContextMenu.h"
```
This capitalization seems suspect

smfr replied:
You should be building on a case sensitive file system.

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -48,6 +49,43 @@
 #if PLATFORM(IOS_FAMILY)
 #import <UIKit/UIColor.h>
 #endif
+@interface WKPDFContextMenuHandler: NSObject {
```
Missing a newline before this, and a space before the :

aprotyas commented on Source/WebCore/platform/LocalizedStrings.cpp:
```diff
@@ -503,6 +503,13 @@ String contextMenuItemTagTranslate(const String& selectedString)
 #endif
 
+#if ENABLE(UNIFIED_PDF)
+String contextMenuItemPDFOpenWith(const String& app)
```
I don't think there's any case when `app != "Preview"`, right? We could just

```cpp
#if ENABLE(UNIFIED_PDF)
String contextMenuItemPDFOpenWithPreview()
{
    return WEB_UI_STRING("Open with Preview", "Open with Preview context menu item");
}
#endif
```

aprotyas replied:
Alternatively if you keep the current definition, you should also format the description argument because it will actually be `Open with {app} context menu item`.

sammygill replied:
Ah yeah good catch! I just ended up creating the string like you demonstrated

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -48,6 +49,43 @@
+    [menu setUserInterfaceLayoutDirection: direction];
+
+    // Open with external PDF viewer
+    menuItem = [[NSMenuItem alloc] initWithTitle:(WebCore::contextMenuItemPDFOpenWith(makeString("Preview"))) action:@selector(openWithPreview) keyEquivalent: @""];
```
Nit, extra space after `keyEquivalent`
```suggestion
    menuItem = [[NSMenuItem alloc] initWithTitle:(WebCore::contextMenuItemPDFOpenWith(makeString("Preview"))) action:@selector(openWithPreview) keyEquivalent:@""];
```

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -48,6 +49,43 @@
+{
+}
+
+- (NSMenu *) createContextMenu: (NSUserInterfaceLayoutDirection) direction
```
Much of the style here is wrong, please consult the WebKit style guide. Mostly you've added a lot of spaces in places.

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -467,6 +501,11 @@
 
+NSMenu* UnifiedPDFPlugin::createContextMenu(const WebMouseEvent&, WebCore::UserInterfaceLayoutDirection uiLayoutDirection)
+{
+    return [m_wkPdfContextMenuHandler createContextMenu: uiLayoutDirection == UserInterfaceLayoutDirection::LTR ? NSUserInterfaceLayoutDirectionLeftToRight : NSUserInterfaceLayoutDirectionRightToLeft];
```
Nit, extra space after `createContextMenu`

```suggestion
    return [m_wkPdfContextMenuHandler createContextMenu:uiLayoutDirection == UserInterfaceLayoutDirection::LTR ? NSUserInterfaceLayoutDirectionLeftToRight : NSUserInterfaceLayoutDirectionRightToLeft];
```

smfr replied:
It's not immediately clear why this function takes an event. I guess we're creating a context menu with items relevant to the event's location? A better name might help.

aprotyas reviewed (commented):
LGTM. Some of the Obj-C styling looks sus

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
```diff
@@ -114,11 +115,13 @@ class PDFPluginBase : public ThreadSafeRefCounted<PDFPluginBase>, public WebCore
     virtual bool isEditingCommandEnabled(StringView commandName) = 0;
 
+    bool handleContextMenuEvent(const WebMouseEvent&);
+    virtual NSMenu* createContextMenu(const WebMouseEvent&, WebCore::UserInterfaceLayoutDirection) = 0;
```
I think at the PDFPluginBase level we should be using platform-agnostic PDFContextMenu, not NSMenu, and PDFPlugin (the old one) should do the conversion internally.

hortont424 replied:
To be even more clear: I don't think UnifiedPDFPlugin should ever have an NSMenu involved (unless there's some good reason I'm not seeing).

smfr replied:
If you do happen to keep `NSMenu* createContextMenu()`, you need to get the refcounting correct. The "create" in the name impies that it returns a retained value. It would be clearer to return a `RetainPtr<NSMenu>`.

sammygill replied:
The main reason was to try and share as much of `handleContextMenuEvent` as possible, but we can certainly keep the `NSMenu` bits into `PDFPlugin` only and have each implementation handle the action somehow

aprotyas replied:
How about put the responsibility of returning a PDFContextMenu on the plugin implementation? That way plugin base stays platform agnostic, PDFPlugin can go from NSMenu to PDFContextMenu internally, and UniversalPDFPlugin can also represent context menus however it wants internally as long as it is providing the plugin base a PDFContextMenu object. (Basically what Tim suggested earlier)

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -550,6 +551,49 @@
+
+    IntPoint point = frameView->contentsToScreen(IntRect(frameView->windowToContents(event.position()), IntSize())).location();
+
+    NSMenu *nsMenu = createContextMenu(event, webPage->userInterfaceLayoutDirection());
```
RetainPtr.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -550,6 +551,49 @@
+            continue;
+
+        // FIXME: This second part of this condition should be removed once UnifiedPDFPlugin::openWithPreview is implemented
+        if ([NSStringFromSelector(item.action) isEqualToString:@"openWithPreview"] && !isUnifiedPDFPlugin())
```
Use `sel_isEqual()` rather than string comparison. But it's a bit weird to compare selectors. Usually NSMenuItems etc have tags and you compare by tag.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
```diff
@@ -550,6 +551,49 @@
+    }
+    PDFContextMenu contextMenu { point, WTFMove(items), WTFMove(openInPreviewIndex) };
+
+    auto sendResult = webPage->sendSync(Messages::WebPageProxy::ShowPDFContextMenu(contextMenu, m_identifier));
```
Does this have to be sync?

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -48,6 +49,43 @@
+    // Open with external PDF viewer
+    menuItem = [[NSMenuItem alloc] initWithTitle:(WebCore::contextMenuItemPDFOpenWith(makeString("Preview"))) action:@selector(openWithPreview) keyEquivalent: @""];
+    [menuItem setTarget: self];
+    [menu addItem:menuItem];
```
Every menuItem is leaked. Use `adoptNS()` with care.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -48,6 +49,43 @@
+    [menu addItem:menuItem];
+    menuItem = (NSMenuItem *)[NSMenuItem separatorItem];
+    [menu addItem:menuItem];
+    return menu;
```
This returns a retained object (which is correct!) but be careful!

smfr commented on Source/WebCore/platform/LocalizedStrings.cpp:
```diff
@@ -503,6 +503,13 @@ String contextMenuItemTagTranslate(const String& selectedString)
 #endif
 
+#if ENABLE(UNIFIED_PDF)
+String contextMenuItemPDFOpenWithPreview()
```
I wish these functions had "Title" in the name; hard to know at the call site that this returns a string.

sammygill replied:
I'll likely do a renaming patch once all the items are added since I'm planning on pulling in the other functions that are defined below

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -410,9 +412,43 @@
+    Vector<PDFContextMenuItem> menuItems;
+
+    // FIXME: We should also set the openInPreviewIndex when UnifiedPdfPlugin::openWithPreview is implemented
+    menuItems.append({ WebCore::contextMenuItemPDFOpenWithPreview(), true, false, 0, true, static_cast<uint8_t>(ContextMenuAction::OpenWithPreview) });
```
Lots of boolean traps here. PDFContextMenuItem needs to use enums more, but this can be fixed in another PR.

aprotyas replied:
+1 to making a wrapper that lets us avoid the bool traps in PDFContextMenu

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -410,9 +412,43 @@
+    auto sendResult = webPage->sendSync(Messages::WebPageProxy::ShowPDFContextMenu(contextMenu, m_identifier));
+    auto [selectedItemTag] = sendResult.takeReplyOr(-1);
+
+    if (selectedItemTag && *selectedItemTag >= 0 && static_cast<unsigned>(*selectedItemTag) < contextMenu.items.size())
```
It's confusing that you use the term "Tag" yet `ContextMenuAction` doesn't have "tag" in the name.

Also I don't think you should assume that tags equate to indexes. Tags might start at 1000 but this code should still work.

sammygill replied:
Ended up renaming to `ContextMenuItemTag`

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -410,9 +412,43 @@
+
+    auto contextMenu = createContextMenu(frameView->contentsToScreen(IntRect(frameView->windowToContents(event.position()), IntSize())).location());
+
+    auto sendResult = webPage->sendSync(Messages::WebPageProxy::ShowPDFContextMenu(contextMenu, m_identifier));
```
Is it possible to asyncify this at some point? Maybe let's put in a FIXME if you think there's enough merit to that

aprotyas reviewed (commented):
LGTM

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -410,9 +412,43 @@
+    auto sendResult = webPage->sendSync(Messages::WebPageProxy::ShowPDFContextMenu(contextMenu, m_identifier));
+    auto [selectedItemTag] = sendResult.takeReplyOr(-1);
+
+    if (selectedItemTag && *selectedItemTag >= 0 && static_cast<unsigned>(*selectedItemTag) < contextMenu.items.size())
```
Nit: You can take advantage of the comparison operators like [`template<class T, class U> operator<=(const optional<T>& opt, const U& value)`](https://en.cppreference.com/w/cpp/utility/optional/operator_cmp) here and re-write this to be a tiny bit more concise.

```suggestion
    if (selectedItemTag >= 0 && selectedItemTag < static_cast<int32_t>(contextMenu.items.size()))
```

sammygill committed:
[UnifiedPDF] Start creating context menu in UnifiedPDFPlugin.
https://bugs.webkit.org/show_bug.cgi?id=265605
rdar://119001947

Reviewed by Simon Fraser.

UnifiedPDF will need to start creating its own context menu so we can
lay the initial groundwork by creating a context menu with a single
item. This will hopefully allow us to easily add additional items in
a subsequent patch without worrying about the architecture too much.
In this patch the item we will use will be the "Open with Preview,"
option.

UnifiedPDFPlugin will create its own PDFContextMenu with items and
send this over to the web page proxy. Each item will contain a tag that
is a ContextMenuAction that will be used to indicate what type of action
to perform once we get the result back.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::contextMenuItemPDFOpenWithPreview):
* Source/WebCore/platform/LocalizedStrings.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::createContextMenu const):
(WebKit::UnifiedPDFPlugin::handleContextMenuEvent):
(WebKit::UnifiedPDFPlugin::performContextMenuAction const):

Canonical link: https://commits.webkit.org/271429@main
